SftpLibsshAdapter/SFTPConnection.cpp

```diff
@@ -0,0 +1,247 @@
+		{
+			throw AuthenticationFailedException(ssh_get_error(m_sshSession));
+		}
+
+		// IMPORTANT only openSSH format private key file are accepted
+		if (ssh_pki_import_privkey_file(privKeyFile.c_str(), privKeyPassPhrase.c_str(), NULL, NULL, &privatekey.sshKey) != SSH_OK)
+		{
+			throw InvalidPrivateKeyException(ssh_get_error(m_sshSession));
+		}
+
+		if (ssh_userauth_publickey(m_sshSession, username.c_str(), privatekey.sshKey) != SSH_AUTH_SUCCESS)
+		{
+			throw AuthenticationFailedPrivateException(ssh_get_error(m_sshSession));
+		}
+	}
+
+	void SFTPConnection::verifySFTPServerIdentity(const std::vector<std::string>& serverFingerPrints)
+	{
+		sshKeyPtr srv_pubkey;
+		hashPtr hash_ptr;
+		hexHashPtr hexHash_ptr;
+
+		if (ssh_get_server_publickey(m_sshSession, &srv_pubkey.sshKey) != SSH_OK)
+		{
+			throw InvalidServerPublicKeyException(ssh_get_error(m_sshSession));
+		}
+			
+		size_t hlen;
+		if (ssh_get_publickey_hash(srv_pubkey.sshKey, SSH_PUBLICKEY_HASH_SHA256, &hash_ptr.hash, &hlen) != SSH_OK)
+		{
+			throw InvalidServerPublicKeyException(ssh_get_error(m_sshSession));
+		}  
+
+		hexHash_ptr.hexHash = ssh_get_hexa(hash_ptr.hash, hlen);
+		if (std::find(serverFingerPrints.begin(), serverFingerPrints.end(), hexHash_ptr.hexHash) == serverFingerPrints.end())
+		{
+			throw ServerFingerPrintNotAcceptedException(hexHash_ptr.hexHash);
+		}
+	}
+
+
+	void SFTPConnection::sftpConnect()
+	{
+		m_sftpSession = sftp_new(m_sshSession);
+		if (!m_sftpSession)
+		{
+			throw SFTPSessionCannotCreateException(ssh_get_error(m_sshSession));
+		}
+
+		if (sftp_init(m_sftpSession) != SSH_OK)
+		{
+			throw SFTPSessionCannotInitException(ssh_get_error(m_sshSession));
+		}
+	}
+
+
+	bool SFTPConnection::isConnected() const
+	{
+		return m_sftpSession;
+	}
+
+
+	void SFTPConnection::upload(const std::string& srcFile, const std::string& dstFile)
+	{
+		if (!isConnected())
+		{
+			throw InvalidConnectionException();
+		}
+
+		// create the destination file in the server
+		sftp_file serverFile = sftp_open(m_sftpSession, dstFile.c_str(), O_WRONLY | O_CREAT | O_TRUNC, DEFAULT_FILE_PERMISSION);
+		if (!serverFile)
+		{
+			throw SFTPCannotCreateRemoteFileException(ssh_get_error(m_sshSession));
+		}
+
+
+		char buffer[MAX_FILE_READ_BUFFER];
+		std::ifstream file(srcFile, std::ios::binary);
+		while (file)
+		{
+			file.read(buffer, sizeof(buffer));  // read the source file content
+			if (file.gcount() > 0)
+			{
+				size_t nwritten = sftp_write(serverFile, buffer, (size_t) file.gcount()); // upload the content on the server
+				if (nwritten != file.gcount())
+				{
+					throw SFTPCannotWriteRemoteFileException(ssh_get_error(m_sshSession));
+				}
+			}
+			else
+			{
+				throw SFTPCannotReadLocalFileException(ssh_get_error(m_sshSession));
+			}
+		}
+
+		sftp_close(serverFile);
+	}
+
+	void SFTPConnection::rename(const std::string& srcFile, const std::string& dstFile)
+	{
+		if (!isConnected())
```

Comment:
for the future, as sftp operations (rename, upload) force to have an active connection may would be a good idea to split down these operations in actual services where they receive as input an active connection either in the constructor or in the operation itself. 

This is a natural way for satisfying a requirement without having to do an if as the requirement is satisfied with the mandatory arguments to execute the operation.